File: memento.java
class memento{
	private static class Memento{
		private String state = null;

		public Memento(String state){
			this.state = state;
		}

		public String getState(){
			return state;
		}
	}

	private static class CareTaker{
		private Memento mementos[] = new Memento[1];

		public int save(Memento mem){
			Memento new_mementos[] = new Memento[mementos.length+1];
			for(int i=0;i<mementos.length;i++){
				new_mementos[i] = mementos[i];
			}
			new_mementos[mementos.length] = mem;
			mementos = new_mementos;
			return (mementos.length-1);
		}

		public Memento restore(int mem_index){
			return mementos[mem_index];
		}
	}

	private static class Originator{
		private Memento mem = null;

		public void setMemento(String state){
			mem = new Memento(state);
			System.out.println("State is set to: "+this.mem.getState());
		}

		public void restoreMemento(Memento new_mem){
			System.out.println("State is reverted back from: "+this.mem.getState()+" to "+new_mem.getState());
			this.mem = new_mem;
		}

		public Memento returnMemento(){
			return this.mem;
		}
	}

	public static void main(String[] args) {
		CareTaker caretaker = new CareTaker();
		Originator originator = new Originator();
		originator.setMemento("State1");
		int index_nr1 = caretaker.save(originator.returnMemento());
		originator.setMemento("State2");
		int index_nr2 = caretaker.save(originator.returnMemento());
		originator.setMemento("State3");
		int index_nr3 = caretaker.save(originator.returnMemento());
		originator.setMemento("State4");
		int index_nr4 = caretaker.save(originator.returnMemento());
		originator.restoreMemento(caretaker.restore(index_nr1));
		originator.restoreMemento(caretaker.restore(index_nr2));
		originator.restoreMemento(caretaker.restore(index_nr3));
		originator.restoreMemento(caretaker.restore(index_nr4));
	}
}

File: memento.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

typedef struct Memento{
	char* state;
} Memento;

struct Memento* init_memento(char* state){
	Memento *memento = malloc(sizeof(Memento));
	memento->state = state;
	return memento;	
}

typedef struct CareTaker {
	int arraysize;
	Memento** mementos;
	int (*save) (struct CareTaker*, struct Memento*);
	Memento* (*restore) (struct CareTaker*, int);
} CareTaker;

int save(CareTaker* caretaker, Memento* memento){
	if(caretaker->arraysize == 0){
		caretaker->mementos = malloc(sizeof(Memento));
		caretaker->mementos[0] = memento;
	} else {
		Memento** mementos_copy = malloc(sizeof(Memento)*(caretaker->arraysize+1));
		for (int i = 0; i < caretaker->arraysize; i++){
			mementos_copy[i] = caretaker->mementos[i];
			caretaker->mementos[i] = NULL;
		}
		mementos_copy[caretaker->arraysize] = memento;
		caretaker->mementos = realloc(caretaker->mementos, (sizeof(Memento)*(caretaker->arraysize+1)));
		for (int i = 0; i < caretaker->arraysize+1; i++){
			caretaker->mementos[i] = mementos_copy[i];
		}
		free(mementos_copy);
	}
	return caretaker->arraysize++;
}

struct Memento* restore(CareTaker* caretaker, int mem_index){
	return caretaker->mementos[mem_index];
}

struct CareTaker* init_caretaker(){
	CareTaker *caretaker = malloc(sizeof(CareTaker));
	caretaker->arraysize = 0;
	caretaker->save = save;
	caretaker->restore = restore;
	return caretaker;
}

typedef struct Orignator {
	Memento* mem;
	void (*setMemento) (struct Orignator*, char* state);
	void (*restoreMemento) (struct Orignator*, Memento* memento);
	Memento* (*returnMemento) (struct Orignator*);
} Orignator;

void setMemento(Orignator* originator, char* state){
	Memento *memento = init_memento(state);
	originator->mem = memento;
	printf("State is set to: %s\n", state);
}

void restoreMemento(Orignator* originator, Memento* memento){
	printf("State is reverted back from: %s to %s\n", originator->mem->state, memento->state);
	originator->mem = memento;
}

Memento* returnMemento(Orignator* originator){
	return originator->mem;
}

struct Orignator* init_originator(){
	Orignator *originator = malloc(sizeof(Orignator));
	originator->setMemento = setMemento;
	originator->restoreMemento = restoreMemento;
	originator->returnMemento = returnMemento;
	return originator;
}

int main(){
	Orignator* originator = init_originator();
	CareTaker* caretaker = init_caretaker();
	originator->setMemento(originator, "State1");
	int ind_1 = caretaker->save(caretaker, originator->returnMemento(originator));
	originator->setMemento(originator, "State2");
	int ind_2 = caretaker->save(caretaker, originator->returnMemento(originator));
	originator->setMemento(originator, "State3");
	int ind_3 = caretaker->save(caretaker, originator->returnMemento(originator));
	originator->setMemento(originator, "State4");
	int ind_4 = caretaker->save(caretaker, originator->returnMemento(originator));
	originator->restoreMemento(originator, caretaker->restore(caretaker,ind_1));
	originator->restoreMemento(originator, caretaker->restore(caretaker,ind_2));
	originator->restoreMemento(originator, caretaker->restore(caretaker,ind_3));
	originator->restoreMemento(originator, caretaker->restore(caretaker,ind_4));
	for (int i = 0; i < caretaker->arraysize; ++i){
		free(caretaker->mementos[i]);
	}
	free(caretaker->mementos);
	free(caretaker);
	free(originator);
	return 0;
}
